src/hotspot/share/utilities/powerOfTwo.hpp

```diff
@@ -40,35 +41,60 @@ constexpr T max_power_of_2() {
   return max_val - (max_val >> 1);
 }
 
-// Returns true iff there exists integer i such that (T(1) << i) == x.
+// Returns true iff there exists integer i such that (T(1) << i) == value.
 template <typename T, ENABLE_IF(std::is_integral<T>::value)>
-constexpr bool is_power_of_2(T x) {
-  return (x > T(0)) && ((x & (x - 1)) == T(0));
+constexpr bool is_power_of_2(T value) {
+  return (value > T(0)) && ((value & (value - 1)) == T(0));
 }
 
-// Log2 of a power of 2
-inline int exact_log2(intptr_t x) {
-  assert(is_power_of_2((uintptr_t)x), "x must be a power of 2: " INTPTR_FORMAT, x);
+// Log2 of a positive, integral value, i.e., largest i such that 2^i <= value
+// Precondition: value > 0
+template<typename T, ENABLE_IF(std::is_integral<T>::value)>
+inline int log2i(T value) {
+  assert(value > T(0), "value must be > 0");
+  const int bits = sizeof(value) * BitsPerByte;
+  return bits - count_leading_zeros(value) - 1;
+}
+
+// Log2 of positive, integral value, i.e., largest i such that 2^i <= value
+// Returns -1 if value is zero
+// For negative values this will return 63 for 64-bit types, 31 for
+// 32-bit types, and so on.
+template<typename T, ENABLE_IF(std::is_integral<T>::value)>
+inline int log2i_graceful(T value) {
+  if (value == 0) {
+    return -1;
+  }
+  const int bits = sizeof(value) * BitsPerByte;
+  return bits - count_leading_zeros(value) - 1;
+}
 
-  const int bits = sizeof x * BitsPerByte;
-  return bits - count_leading_zeros(x) - 1;
+// Log2 of a power of 2, i.e., i such that 2^i == value
+// Preconditions: value > 0, value is a power of two
+template<typename T, ENABLE_IF(std::is_integral<T>::value)>
+inline int log2i_exact(T value) {
+  assert(is_power_of_2(value),
+                       "value must be a power of 2: " UINT64_FORMAT,
+                       (uint64_t)static_cast<typename std::make_unsigned<T>::type>(value));
```

Comment:
This looks strange and complicated for something that ought to be straightforward.  There was a comment in conversation suggesting this is working around some problem with some test.  Please explain, because I'd really like this to go back to just a cast to uint64_t if at all possible.  Sorry I didn't bring this up earlier.

Reply:
For signed int 0x80000000 casting to uint64_t will change the value to 0xFFFFFFFF80000000, which is avoided by making the value unsigned before casting  (0x0000000080000000). After we changed the precondition to require value to be positive this is only an issue for what the assert prints in this edge case, so I agree we can simplify it.